### PR TITLE
Refactor(cv_deploy): Improve manifest sync code

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Refactor manifest sync code

## Component(s) name

pyavd

## Proposed changes

**Changes**

- Refactors Static Configuration Studio container sync to build a single container/root plan before applying changes.
- Treats Studio root assignments as parent edges in the same graph as container child assignments.
- Adds an internal existing-state model and container plan model to make validation, diffing, and apply steps clearer.
- Validates existing container state before mutating CloudVision, including missing references, multiple parents, manual orphan containers, and manifest-managed containers rehomed under uncontrolled manual parents.
- Preserves recovery behavior for manifest-managed orphan containers so stale manifest-owned state can still be cleaned up or reparented.
- Uses the computed container plan for configlet deletion validation, allowing stale configlets to be deleted only when all holders are rewritten/deleted or otherwise controlled by the deploy.
- Consolidates Studio root reconciliation into the container plan while preserving existing documented root ordering behavior.
- Updates user-facing terminology from AVD-managed to manifest-managed where applicable.
- Adds docstrings and targeted comments for the graph model, root handling, validation rules, and manual subtree deletion behavior.

## How to test

- Existing tests are functionally untouched but improved a few comments.
- Added a few extra tests for corner cases.

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/extensions/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 1]
```diff
@@ -22,13 +23,66 @@
...
+    def touched_container_ids(self) -> set[str]:
+        return {container.id for container in self.containers_to_push} | self.container_ids_to_delete
 
 
 async def deploy_static_config_studio_manifest_to_cv(manifest: AvdManifest, deployment_result: DeployToCvResult, cv_client: CVClient) -> None:
     """
     Deploy a manifest (configlets/containers) to CloudVision using the "Static Configuration" Studio.
 
-    TODO: Implement strict mode to remove any containers/configlets not managed by AVD from the Studio.
+    TODO: Implement strict mode to remove any containers/configlets not managed by the manifest from the Studio.
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:86]: The last TODO below is outdated.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 2]
```diff
@@ -48,57 +102,106 @@ async def deploy_static_config_studio_manifest_to_cv(manifest: AvdManifest, depl
...
+        containers_by_id=containers_by_id,
+        children_by_parent_id=children_by_parent_id,
+        parent_ids_by_child_id=dict(parent_ids_by_child_id),
+        manifest_managed_container_ids=manifest_managed_container_ids,
+    )
+
+
+def _build_container_plan(cv_manifest: CVManifest, existing_state: _ExistingContainerState) -> _ContainerPlan:
+    """
+    Build the desired container/root diff from a trusted existing state.
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py]: "from a trusted existing state" seems wrong. We call `_validate_existing_container_state` after this function.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 3]
```diff
@@ -113,98 +216,169 @@ async def _sync_containers(
...
+            ]
+            violations.extend(
+                f"Manifest-managed container '{container.display_name}' (id={container_id}) is currently a child of {existing_state.parent_display(parent_id)}"
+                for parent_id in manual_parent_ids
+            )
             continue
 
-        # Parent is out of our control.
-        violations.append((cast("str", parent.display_name), parent_id, child_name, child_id))
+        if not parent_ids_set and container_id not in container_ids_to_delete:
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py]: The issues below could've been created by cv_deploy itself before introducing the fix in https://github.com/aristanetworks/avd/pull/7024. API calls are required to resolve these issues, which is not super user friendly.
```
"errors": ["The existing Static Configuration Studio container state is inconsistent. Resolve these issues and re-run: Manual container 'MANUAL_SUB_CONTAINER' (id=44f0b5b4-8343-45e1-befb-4c8375609860) is orphaned; Manual container 'MANUAL_SUB_CONTAINER' (id=7644c942-de16-4819-a760-7b14f343f1d2) is orphaned; Manual container 'MANUAL_SUB_CONTAINER' (id=dbb5714e-e946-49d8-9c14-534adee4211b) is orphaned"]
```

Also, that means if a user has any lingering manual container without a parent and without being in the Studio root list, the next cv_deploy run will start failing where it previously succeeded. It feels a bit breaking no?

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 4]
```diff
@@ -22,13 +23,66 @@
 
 
 STATIC_CONFIGURATION_STUDIO_ID = "studio-static-configlet"
+STUDIO_ROOT_PARENT_ID = "<Static Configuration Studio roots>"
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:26]: This is not a "real" ID correct? Let's add a small docstring to make it clear that it is a sentinel.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 5]
```diff
@@ -113,98 +216,169 @@ async def _sync_containers(
...
+    return _ContainerPlan(
+        containers_to_push=containers_to_push,
+        containers_to_skip=containers_to_skip,
+        containers_to_delete=containers_to_delete,
+        new_root_ids=new_root_ids,
+        root_ids_changed=new_root_ids != existing_state.root_ids,
+    )
+
+
+def _validate_existing_container_state(existing_state: _ExistingContainerState, container_plan: _ContainerPlan) -> None:
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:254]: We need to add unit tests to cover the new violations.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 6]
```diff
@@ -113,98 +216,169 @@ async def _sync_containers(
...
+            studio_id=STATIC_CONFIGURATION_STUDIO_ID,
+            workspace_id=workspace_id,
+            input_path=["configletAssignmentRoots"],
+            inputs=container_plan.new_root_ids,
+        )
+    else:
+        LOGGER.info("deploy_static_config_studio_manifest_to_cv: Studio root container assignments are already in the desired state.")
+
+
+def _raise_configlet_holder_violations(violations: list[tuple[str, str, str, str]]) -> None:
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:331]:
```suggestion
def _raise_if_configlet_holder_violations(violations: list[tuple[str, str, str, str]]) -> None:
```

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 5 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:254]: I will add them.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 3 above)
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py]: When writing unit tests, I noticed that `and container_id not in container_ids_to_delete` is unreachable in practice because `add_container_subtree_to_delete` is only seeded from 1/ children dropped by a pushed managed parent and 2/ managed containers not in manifest. A manual container with no parent fits neither path, so it can never be in `containers_to_delete`.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 7]
```diff
@@ -22,13 +23,66 @@
...
+    root_ids: list[str]
+    containers_by_id: dict[str, ConfigletAssignment]
+    children_by_parent_id: dict[str, list[str]]
+    parent_ids_by_child_id: dict[str, list[str]]
+    manifest_managed_container_ids: set[str]
+
+    def container_name(self, container_id: str) -> str:
+        container = self.containers_by_id.get(container_id)
+        if container is None:
+            return container_id
```
Comment (carl-baillargeon) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:49]: This is dead cause because this is only reachable through `parent_display` when a parent_id is missing from `containers_by_id` but since `parent_ids_by_child_id` is populated from existing containers children lists, the parent_id is always a real container ID.

Comment (ClausHolbechArista) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:49]: typechecker doesn't know that.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 6 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:331]: I think the current is more clear.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 7 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:49]: Also this is not unreachable. It is just not exercised in our current production code.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 3 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py]: Removed most of the violations.

**Review by carl-baillargeon — APPROVED** [review]
LGTM if internal CI passes: https://github.com/aristanetworks/avd-internal/actions/runs/26592062231

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 8]
```diff
@@ -22,16 +23,68 @@
...
+    Existing Static Configuration Studio container graph.
+
+    Studio roots are modeled as parent edges from STUDIO_ROOT_PARENT_ID, so parent validation can use the
+    same logic for root assignments and regular container child assignments.
+    """
+
+    containers: list[ConfigletAssignment]
+    root_ids: list[str]
+    containers_by_id: dict[str, ConfigletAssignment]
+    children_by_parent_id: dict[str, list[str]]
```
Comment (alexeygorbunov) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:42]: Based on the type definition - should it be `children_ids_by_parent_id`? Otherwise it feels like list[ConfigletAssignment]

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 9]
```diff
@@ -22,16 +23,68 @@
...
+    """
+    Planned container and root mutations.
+
+    containers_to_delete can include manual containers, but only when they are below a manifest-managed branch
+    that this deploy updates/deletes and would otherwise leave them orphaned in CV.
+    """
+
+    containers_to_push: list[CVContainer]
+    containers_to_skip: list[CVContainer]
+    containers_to_delete: dict[str, str]
```
Comment (alexeygorbunov) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:69]: Would docstring be helpful here for undrstanding what are keys and values? id:name?

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py [hunk 10]
```diff
@@ -113,98 +215,146 @@ async def _sync_containers(
...
+        container = existing_state.containers_by_id[container_id]
+        parent_ids = existing_state.parent_ids_by_child_id.get(container_id, [])
+        parent_ids_set = set(parent_ids)
+        manual_parent_ids = [
+            parent_id
+            for parent_id in parent_ids_set
+            if parent_id != STUDIO_ROOT_PARENT_ID and not parent_id.startswith(AVD_ENTITY_PREFIX) and parent_id not in container_ids_to_delete
+        ]
+        violations.extend(
+            f"Manifest-managed container '{container.display_name}' (id={container_id}) is currently a child of {existing_state.parent_display(parent_id)}"
```
Comment (alexeygorbunov) [inline python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:275]: Is this message clear for the user to understand what the issue/inconsistency is? Previous version of the message was clear about parent not being managed by the manifest.

**Review by alexeygorbunov — APPROVED** [review]
Left few minor comments (still ok if those are not addressed).
LGTM

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 8 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:42]: Fixed in https://github.com/aristanetworks/avd/pull/7040.

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 9 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:69]: Tracked here: https://github.com/aristanetworks/avd-internal/issues/579

File: python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py (hunk 10 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_cv/workflows/deploy_static_config_studio_manifest_to_cv.py:275]: Will be covered in https://github.com/aristanetworks/avd/pull/7040.
